File: src/controllers/SearchController.js
/**
 * src/controllers/SearchController.js
 *
 * Búsqueda universal: recorre todas las secciones en Drive
 * y filtra por nombre, sección, subcarpeta y tipo de archivo.
 *
 * Query params:
 *   q         → texto libre (nombre de documento)
 *   section   → id de sección  (ej: "actividades")
 *   subfolder → nombre de subcarpeta
 *   type      → tipo de archivo (ej: "PDF", "DOCX")
 */

const driveService           = require("../services/DriveService");
const { SECTIONS }           = require("../config/sections");

const SearchController = {

  search: async (req, res, next) => {
    try {
      const { q, section, subfolder, type } = req.query;

      const hasQuery = q || section || subfolder || type;

      let results   = [];
      let allTypes  = [];
      let allFolders = [];

      if (hasQuery) {
        results = await driveService.searchAcrossSections(SECTIONS, q, {
          sectionId: section  || null,
          subfolder: subfolder || null,
          fileType:  type      || null,
        });

        // Calcular listas únicas para los filtros desplegables
        allTypes   = [...new Set(results.map(d => d.typeLabel))].sort();
        allFolders = [...new Set(results.map(d => d.folder))].sort();
      }

      res.render("search/results", {
        title:       q ? `Resultados: "${q}"` : "Buscar documentos",
        query:       q       || "",
        filterSection:   section   || "",
        filterSubfolder: subfolder || "",
        filterType:      type      || "",
        results,
        total:       results.length,
        sections:    SECTIONS,
        allTypes,
        allFolders,
        hasQuery:    !!hasQuery,
      });
    } catch (err) { next(err); }
  },
};

module.exports = { SearchController };

File: Controllers/SearchController.cs
// Controllers/SearchController.cs
using Microsoft.AspNetCore.Mvc;
using bilbioteca_virtual_aca.Models;
using bilbioteca_virtual_aca.Models.ViewModels;
using bilbioteca_virtual_aca.Services;

namespace bilbioteca_virtual_aca.Controllers;

public class SearchController : Controller
{
    private readonly DriveService _drive;

    public SearchController(DriveService drive)
    {
        _drive = drive;
    }

    // ── GET /Search?q=...&section=...&subfolder=...&type=... ──
    public async Task<IActionResult> Index(
        string? q,
        string? section,
        string? subfolder,
        string? type)
    {
        ViewData["Title"]       = string.IsNullOrEmpty(q) ? "Buscar documentos" : $"Resultados: \"{q}\"";
        ViewBag.SearchQuery     = q;

        var hasQuery = !string.IsNullOrEmpty(q) || !string.IsNullOrEmpty(section)
                    || !string.IsNullOrEmpty(subfolder) || !string.IsNullOrEmpty(type);

        var model = new DriveSearchViewModel
        {
            Query           = q        ?? "",
            FilterSection   = section  ?? "",
            FilterSubfolder = subfolder ?? "",
            FilterType      = type     ?? "",
            Sections        = Sections.All,
            HasQuery        = hasQuery,
        };

        if (hasQuery)
        {
            try
            {
                model.Results = await _drive.SearchAsync(
                    Sections.All, q, section, subfolder, type);

                model.Total      = model.Results.Count;
                model.AllTypes   = model.Results.Select(r => r.TypeLabel).Distinct().OrderBy(t => t).ToList();
                model.AllFolders = model.Results
                    .Where(r => r.FolderName is not null)
                    .Select(r => r.FolderName!)
                    .Distinct().OrderBy(f => f).ToList();
            }
            catch (Exception ex)
            {
                ViewBag.SearchError = ex.Message;
            }
        }

        return View(model);
    }
}
